src/OhmmsPETE/OhmmsMatrix.h

```diff
@@ -107,38 +107,32 @@ class Matrix
     X.attachReference(ref, TotSize);
   }
 
+  template<typename Allocator = Alloc, typename = IsHostSafe<Allocator>>
   inline void add(size_type n) // you can add rows: adding columns are forbidden
   {
-    static_assert(MemType == MemorySpace::HOST, "Matrix::add MemType must be MemorySpace::HOST");
     X.insert(X.end(), n * D2, T());
     D1 += n;
   }
 
+  template<typename Allocator = Alloc, typename = IsHostSafe<Allocator>>
   inline void copy(const This_t& rhs)
   {
-    static_assert(MemType == MemorySpace::HOST, "Matrix::copy MemType must be MemorySpace::HOST");
     resize(rhs.D1, rhs.D2);
     assign(*this, rhs);
   }
 
   // Assignment Operators
   inline This_t& operator=(const This_t& rhs)
   {
-    static_assert(MemType == MemorySpace::HOST, "Matrix::operator= MemType must be MemorySpace::HOST");
     resize(rhs.D1, rhs.D2);
-    return assign(*this, rhs);
-  }
-
-  inline const This_t& operator=(const This_t& rhs) const
-  {
-    static_assert(MemType == MemorySpace::HOST, "Matrix::operator= MemType must be MemorySpace::HOST");
-    return assign(*this, rhs);
+    if (allocator_traits<Alloc>::is_host_accessible)
```

Comment:
copy constructor and assign operator are special. Because of the compiler generated versions, I cannot make SFINAE working as intended and it caused problems. So I'm leaving if-statement here, in fact we can have if constexpr in the future to kill any overhead.